adapters/validation.py

```diff
@@ -0,0 +1,213 @@
+                method = getattr(cls, attr)
+                if callable(method):
+                    validators.append(method)
+        node_kwargs['validators'] = validators
+
+        cls._node = ValidatorListNode(**node_kwargs)
+        return cls
+
+    def __call__(cls):
+        raise TypeError("can't instantiate {}".format(cls.__name__))
+
+    def __hash__(cls):
+        if cls is DeclarativeValidationNode:
+            return super().__hash__()
+        else:
+            return hash(cls._node)
+
+    def __eq__(cls, other):
+        if cls is other:
+            return True
+        elif cls is DeclarativeValidationNode:
+            return False
+        else:
+            return cls._node == other
+
+    def __getattr__(cls, name):
+        if name in cls.NODE_ATTRS:
+            return getattr(cls._node, name)
+        else:
+            return super().__getattr__(name)
+
+
+class DeclarativeValidationNode(object, metaclass=DeclarativeValidationNodeMeta):
+    pass
+
+
+class ValidationResult(object):
+    def __init__(self, tree, data, errors=None, failed_nodes=frozenset()):
+        self.tree = tree
+        self.data = data
+        if errors is None:
+            self.errors = []
+        else:
+            self.errors = errors
+        self.failed_nodes = frozenset(failed_nodes)
+
+    @property
+    def is_valid(self):
+        return not self.errors
+
+    def fields_with_errors(self):
+        fields = set()
+        for error in self.errors:
+            fields.add(error.field)
+        return fields
+
+
+class ValidationTree(object):
+    def __init__(self, final_node):
+        topological_order = [final_node]
+        all_nodes = {final_node}
+        nodes = {final_node}
+        while nodes:
+            node = nodes.pop()
+            new_nodes = node.dependencies - all_nodes
+            all_nodes.update(new_nodes)
+            nodes.update(new_nodes)
+            topological_order.extend(new_nodes)
+        topological_order.reverse()
+
+        nodes = {final_node}
+        for i in range(len(all_nodes) + 1):
+            node = nodes.pop()
+            nodes.update(node.dependencies)
+            if not nodes:
+                break
+        else:
+            raise CycleError('cycle in validators')
+
+        self._final_node = final_node
+        self._all_nodes = all_nodes
+        self._topological_order = topological_order
+
+    def validate(self, data, **kwargs):
+        failed_nodes = set()
+        errors = []
+        for node in self._topological_order:
+            if any(d in failed_nodes for d in node.dependencies):
+                failed_nodes.add(node)
+            else:
+                try:
+                    new_data = node.validate(data, **kwargs)
+                except ValidationError as e:
+                    failed_nodes.add(node)
+                    errors.append(e)
+                else:
+                    if new_data is not None:
+                        data = new_data
```

Comment:
I didn't see any tests with validators that `return`, so I suspect we're not covering this line. Do we want to be recommending a mutation api?